swan/modeller/base_modeller.py

```diff
@@ -0,0 +1,65 @@
+
+import abc
+from typing import Optional, Tuple, TypeVar
+
+from ..dataset.swan_data_base import SwanDataBase
+from ..state import StateH5
+from ..type_hints import PathLike
+
+T = TypeVar('T')
+
+
+class BaseModeller:
+    """Base class for the modellers."""
+
+    def __init__(self, data: SwanDataBase) -> None:
+        self.state = StateH5()
+        self.smiles = data.dataframe.smiles.to_numpy()
+
+    @abc.abstractmethod
+    def train_model(self, frac: Tuple[float, float] = (0.8, 0.2), **kwargs):
+        """Train the model using the given data.
+
+        Parameters
+        ----------
+        frac
+            fraction to divide the dataset, by default [0.8, 0.2]
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def validate_model(self) -> Tuple[T, T]:
```

Comment:
```suggestion
    def validate_model(self) -> Tuple[Any, Any]:
```
FYI, this is not exactly an appropriate place to use typevars, as there is no other `T`-annotated class/method parameter that can be used for inferring its type.

Reply:
I think it is the right place to have a constraint because, in that case in particular `T = T` but I need to specify a constrained version of T like:
```py
import numpy as np
import torch

T = TypeVar('T', np.ndarray, torch.Tensor)

def validate_model(self) ->Tuple[T, T]:
```

Reply:
You are right and I still don't get why a get a warning. Anyhow, I am changing the type to Any

Reply:
Since we're dealing with an abstract baseclass here, there _is_ a way of making this work: by making the class parametrizable:
``` python
from typing import Union, TypeVar, Tuple, Generic, TYPE_CHECKING

import numpy as np
import torch

# `bound` preserves all sub-type information, which might be useful
T = TypeVar('T', bound=Union[np.ndarray, torch.Tensor])

class BaseModeller(Generic[T]):
    def validate_model(self) -> Tuple[T, T]: ...

# Example subclasses
class NDArrayModeller(BaseModeller[np.ndarray]): ...
class TensorModeller(BaseModeller[torch.Tensor]): ...
```
In this example `T` is variable in the ABC (and the class is thus subscriptable) while its value is fixed in the example subclasses.

Reply:
It works! I am just a bit disappointed that we need to write C++ template metaprogramming in the Types to provide a very straightforward type

Reply:
Yeah... It would not be an understatement to describe static typing as its own mini-language at this point.